tests/test_regex.py

```
@@ -74,5 +300,37 @@ def test_named_groups(self):
     def test_begining(self):
         assert_can_generate('^abc')
 
+    def test_caret_in_the_middle_does_not_generate_anything(self):
+        r = re.compile('a^b')
+
+        with pytest.raises(he.NoSuchExample):
+            h.find(regex(r), r.match)
```

Does this not generate anything at all, in which case you can get the exception from `regex(r).example()`?  Or does it not generate any matches, in which case it should detect that and raise `InvalidArgument`?

This is the everything-matches invariant again - it's fine to not support some valid regexes, but they should raise InvalidArgument instead of exploding later.

---

You probably out of context here. Thing is that this regex "a^b" should not match anything. You can try that. But the way generator works currently it would generate strings "ab", which is incorrect. So I added this test and update generation to prevent generation in that case.

---

1. `'a^b'` does not match any string.  Thanks to the filter, the strategy therefore generates no examples.  (so far, so goo)
2. Using `r.match` as the predicate is therefore redundant.  
3. `lambda s: True` would be more efficient, and clearly does not exclude possible outputs.
4. Using `regex(r).example()` is more efficient again - it basically uses `find` internally and stops at the first example, where `find` keeps trying until it finds the simplest example.